### PR TITLE
Semantic zooming

### PR DESCRIPTION
**Pull request by JaspervanTilburg**

### CODE REVIEW

File: src/main/java/gui/GraphDrawer.java [hunk 1]
```diff
@@ -233,10 +245,45 @@ private boolean inView(double[] coordinates) {
...
+        if (!node.getChildren().isEmpty()) {
+            SequenceNode child = graph.getNode(node.getChild(0));
+            if (child.isSNP() && child.isCollapsed()) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+//    public void drawNodes() {
```
Comment (LBoleij) [inline src/main/java/gui/GraphDrawer.java]: Can you remove this commented code?

File: src/main/java/gui/GraphDrawer.java [hunk 2]
```diff
@@ -233,10 +245,45 @@ private boolean inView(double[] coordinates) {
         coordinates[2] = width;
         coordinates[3] = height;
 
+        this.coordinates.put(node.getId(), coordinates);
         return coordinates;
 
     }
 
+    private boolean relativeWidth(SequenceNode node) {
```
Comment (LBoleij) [inline src/main/java/gui/GraphDrawer.java]: add some JavaDoc

File: src/main/java/gui/GraphDrawer.java [hunk 3]
```diff
@@ -324,25 +377,62 @@ private void drawAnnotations(SequenceNode node, double[] coordinates) {
...
+            return false;
+        }
+        upperNode.setSNP(true);
+        upperNode.setCollapsed(true);
+        lowerNode.setSNP(true);
+        lowerNode.setCollapsed(true);
+        return true;
+    }
+
+    private void checkExtremeNode(SequenceNode node) {
```
Comment (LBoleij) [inline src/main/java/gui/GraphDrawer.java]: Add JavaDoc

File: src/main/java/gui/GraphDrawer.java [hunk 4]
```diff
@@ -355,10 +445,33 @@ private void drawNode(SequenceNode node) {
         }
     }
 
+    private void drawSNPBubble(SequenceNode upperNode, SequenceNode lowerNode) {
```
Comment (LBoleij) [inline src/main/java/gui/GraphDrawer.java]: Add JavaDoc

**Comment by JaspervanTilburg** [comment]
Javadoc is added and checkstyle errors are resolved

File: src/main/java/gui/GraphDrawer.java [hunk 5]
```diff
@@ -324,163 +389,300 @@ private void drawAnnotations(SequenceNode node, double[] coordinates) {
...
-    private void checkExtremeNode(SequenceNode node) {
-        double[] coordinates = getCoordinates(node);
-        if (coordinates[0] <= 0 && coordinates[0] + coordinates[2] > 0) {
-            mostLeftNode = node;
+    /**
+     * Check if a column can be collapsed into a SNP bubble. If so set the nodes isSNP attribute to true.
+     *
+     * @param columnID The column to check for a SNP bubble
+     */
+    private void checkSNPBubble(int columnID) {
```
Comment (Jipje) [inline src/main/java/gui/GraphDrawer.java]: Looks good

File: src/main/java/gui/GraphDrawer.java [hunk 6]
```diff
@@ -324,163 +389,300 @@ private void drawAnnotations(SequenceNode node, double[] coordinates) {
...
+            double midY = upperCoordinates[Y_INDEX] + upperCoordinates[HEIGHT_INDEX] / 2;
+            double upY = midY - upperCoordinates[HEIGHT_INDEX];
+            double downY = midY + upperCoordinates[HEIGHT_INDEX];
+
+            if (upperNode.isHighlighted()) {
+                gc.setLineWidth(LINE_WIDTH);
+                gc.strokePolygon(new double[]{leftX, midX, rightX}, new double[]{midY, upY, midY}, POLYGON_POINTS);
+                gc.strokePolygon(new double[]{leftX, midX, rightX}, new double[]{midY, downY, midY}, POLYGON_POINTS);
+            }
+            gc.setFill(Color.CHOCOLATE);
```
Comment (Jipje) [inline src/main/java/gui/GraphDrawer.java]: Chocolate and brown? They look very similair...

**Comment by LBoleij** [comment]
Ik vind mooi

File: src/main/java/gui/GraphDrawer.java (hunk 6 above)
Comment (JaspervanTilburg) [inline src/main/java/gui/GraphDrawer.java]: What colours do you suggest?

Comment (LBoleij) [inline src/main/java/gui/GraphDrawer.java]: Paars met goud?

Comment (LBoleij) [inline src/main/java/gui/GraphDrawer.java]: We decided to keep Chocolate.

**Pull request merged**